### DISCUSSION TITLE
TypeScript Mismatch using Pick or Omit

### DISCUSSION THREAD

**Discussion by mohrazzak** [colinhacks/zod#3117] [Q&A] [answered]
I have a NestJS application using Prisma and Zod, where I define a `StoreOwnerModel` representing a database table with all fields, and a `StoreOwnerEntity` which is a minimal representation containing only specific fields for frontend responses.

I have the following function to retrieve a single `StoreOwnerEntity` based on the `storeOwnerId`:

```typescript
async getOne(storeOwnerId: number): Promise<StoreOwnerEntity> {
  const storeOwnerOfTypeModel = await this.prismaService.storeOwner.findUniqueOrThrow({
    where: { id: storeOwnerId },
  });

  // My question is why TypeScript is not throwing an error here, even though
  // I'm trying to return a `StoreOwnerModel` instead of the expected `StoreOwnerEntity`.
  // Shouldn't there be a type mismatch error since the return type is declared as `StoreOwnerEntity`?

  // return storeOwnerOfTypeModel; // No TypeScript Error
  return this.storeOwnerEntityMapper.modelToEntity(storeOwnerOfTypeModel);
}
```

Additionally, I've defined a `StoreOwnerEntitySchema` using Zod with only the necessary fields:

```typescript
export const StoreOwnerEntitySchema = StoreOwnerSchema.pick({
  id: true,
  email: true,
});

export class StoreOwnerEntity extends createZodDto(StoreOwnerEntitySchema) {}
```

```typescript
export const StoreOwnerSchema = z.object({
  id: z.number().int(),
  email:  z.string().email();,
  createdAt: z.date().default(new Date()),
});
```

I would appreciate insights into why TypeScript is not throwing an error in this case and guidance on ensuring type safety when mapping and returning entities. Thank you!

**Comment by JacobWeisenburger** [comment, marked as answer]
Please send a full reproducible code example.

![image](https://github.com/colinhacks/zod/assets/31667350/56b9ecb4-07ca-47a6-8142-36c0d87aed94)
